### PR TITLE
feature_630_UIUC_GDF_TDF

### LINKED ISSUE #630: Add S2S use case for TC Genesis Density Function (GDF) and TC Track Density Function (TDF)
## Describe the New Feature ##
Add a use case demonstrating the computation and visualization of Genesis Density Function (GDF). This metric is used in S2S application to visualize the distribution and frequency of TC genesis events across a user-defined time period. 

### Acceptance Testing ###
Forecast data: GFDL vortex tracker output in "genesis mode" from NWP model forecast data
Observation data: Global ATCF bdecks data
MET tool: tc_gen
METplus wrapper: tc_gen_wrapper.py

### Time Estimate ###
2-4 days of work

### Funding Source ###
~~UIUC S2S 7790971~~
2785051

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [ ] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##

dtcenter/MET#1430

Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by DanielAdriaansen**
Closes #630.

- [x] Describe testing already performed for these changes:</br>
I did testing on kiowa, with test data that I provided to George and he staged for me. I tested using the MET develop branch, which is the version that contains changes in dtcenter/MET#1714 needed for this use case. I am using an Anaconda Python environment, called `s2stc`, located at: `/home/dadriaan/.conda/envs/s2stc`.

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
We should run the entire automation suite, since it looks like my latest push ran the documentation portion only.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [ ] Do these changes include sufficient testing updates? **[Yes or No]**
I did not write any new tests for this use case.

- [ ] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
I don't know.

## Pull Request Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, and **Milestone**
- [x] After submitting the PR, select **Linked Issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Review by georgemccabe — CHANGES_REQUESTED** [review]
The changes look mostly good. The branch did not have the latest changes from last pull request so there were differences in the output. I merged develop into the branch and the tests are running now.

The use case documentation results in a page that has SyntaxError as the title:
https://metplus.readthedocs.io/en/feature_630_uiuc_gdf_tdf/generated/model_applications/s2s/TCGen_fcstGFSO_obsBDECKS_GDF_TDF.html#sphx-glr-generated-model-applications-s2s-tcgen-fcstgfso-obsbdecks-gdf-tdf-py

This should be fixed before approving.

**Comment by DanielAdriaansen** [comment]
> The changes look mostly good. The branch did not have the latest changes from last pull request so there were differences in the output. I merged develop into the branch and the tests are running now.

Interesting. I wonder if this is because the last PR was done after I did the pull, merge, and this PR creation? Or did I miss something when generating the PR?

> 
> The use case documentation results in a page that has SyntaxError as the title:
> https://metplus.readthedocs.io/en/feature_630_uiuc_gdf_tdf/generated/model_applications/s2s/TCGen_fcstGFSO_obsBDECKS_GDF_TDF.html#sphx-glr-generated-model-applications-s2s-tcgen-fcstgfso-obsbdecks-gdf-tdf-py
> 
> This should be fixed before approving.

I just fixed the docs issue, thanks for catching it! I see there are other errors from automation and I will check on those tomorrow.

**Comment by georgemccabe** [comment]
Thanks for making those changes, @DanielAdriaansen . The missing changes from develop were merged in yesterday and likely after you had created this pull request. This is common when many pull requests are being created/approved around the same time. The updates look good to me. I can't seem to approve the PR without you asking for another review, but I do approve.

Regarding differences in the test output, there should be no differences in the existing output, but there should be 1 failure in the checks for the difference test in the new use case group. This happens because there is no truth data generated yet. The failure should only occur in the new use case for the pull request run only. This is what I see, so we are good.

After this PR is merged, we will need to update the develop-ref branch from develop to regenerate the truth data to include the new output. This is usually done with a pull request to easily track when the truth data changed. Almost every time the PR shows conflicts, but they resolution is obvious (use the incoming aka develop version). 

We should also either move the s2s:4:NEW use case into another s2s use case group or remove :NEW from the end and make it its own group. This decision depends on the size of the output artifacts from these groups and the time it takes to run the use cases. We want to keep the size of the output artifacts relatively small so they can be easily downloaded if needed. We also want to keep the total run time of the use case groups low since they can run in parallel and a long execution time can slow down the total run time of the workflow. Keep in mind the run time listed on the summary diagram for a workflow includes setup/tear down of the environment. Also, installing additional python packages may also slow things down (especially if it uses conda), but we are eventually rethinking how we obtain these packages to reduce this burden.

**Pull request merged**
